### ISSUE TITLE
[GNOME Linux] No window decoration (shadow) if a window title bar style is custom

### ISSUE THREAD

**Issue by proninyaroslav** [microsoft/vscode#225073]
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.92.0
- OS Version: Fedora 40 (GNOME 46)

Steps to Reproduce:

1. Set `window.titleBarStyle` options to `custom` in the settings
2. Run VSCode in the Wayland mode, for example, type it in the terminal:
```
$ ELECTRON_OZONE_PLATFORM_HINT=auto code
```

**Expected behaviour**

There should be a window decoration (shadow) as it is in XWayland compatibility mode:
![image](https://github.com/user-attachments/assets/05cc418e-473a-447a-9176-b6ce88ac2ad8)

**Actual behaviour**

There is no window decoration/shadow:
![image](https://github.com/user-attachments/assets/ab49b4c1-9cf5-4d51-8f94-0500ec73bd32)

But If I set `window.titleBarStyle` to `native`, the window decoration/shadow appears:
![image](https://github.com/user-attachments/assets/d56b4c6d-b2e9-4cde-9e0c-ca6f079555e6)

**Comment by proninyaroslav** [comment]
Tested on `1.92.1`, still has problem.

**Comment by proninyaroslav** [comment]
I noticed another fact: in Wayland mode, a new window opens as `Visual Studio Code - URL Handler`, and not as it should be (`Visual Studio Code`). Maybe this is the reason.

**Comment by cbrnr** [comment]
I can confirm the missing window decorations (i.e. shadows) when running on Wayland when using custom title bar style. It would be really nice if this worked, because

1. the native title bar is not an option for me (the pane shortcut buttons are missing and it takes up more space), and
2. running via XWayland is also not an option, because I use fractional scaling (150%), which is very blurry in XWayland (but super crisp on Wayland).

Regarding the comment on "Visual Studio Code - URL Handler", I still haven't figured out why there are two desktop files, but you can at least fix the issue with two task bar entries (and/or missing icons) by setting `StartupWMClass=code-url-handler` in `~/.local/share/applications/code.desktop` (create a copy of `/usr/share/applications/code.desktop` in that location and edit the copy in your home folder), see https://github.com/microsoft/vscode/issues/129953#issuecomment-1315491384 and https://github.com/microsoft/vscode/issues/129953#issuecomment-2364167906.

**Comment by proninyaroslav** [comment]
@cbrnr 
Yes, I was a little confused about URL Handler. I eventually found that even with the system window frame, VSCode still launched as URL Handler.

**Comment by proninyaroslav** [comment]
@jhasse 
Well, as a workaround, we can add a border to at least somehow distinguish the window frame from the rest.

**Comment by cbrnr** [comment]
However, Google Chrome does have borders and a shadow even though it is the basis of Electron (and it also has a custom title bar). So it might be worth looking into how they achieve this.

**Comment by proninyaroslav** [comment]
@cbrnr 
I confirm, I'm using a custom frame in Chrome and it has a shadow.

**Comment by deepak1556** [comment]
The shadow with native titlebar are from the CSD implementation in Electron. For the custom titlebar Electron shares the implementation with Chrome and based on https://issues.chromium.org/issues/40158897 there is no work actively being tracked in upstream. Closing as not actionable on our end.

**Comment by cbrnr** [comment]
@deepak1556 but wait, as I've mentioned previously, Chromium does have shadows when using a custom title at. Why not VS Code then?

**Comment by deepak1556** [comment]
@cbrnr please refer to https://issues.chromium.org/issues/40158897#comment4

**Comment by cbrnr** [comment]
I don't understand. This issue is 3 years old and marked as obsolete, probably because it already works?

**Comment by proninyaroslav** [comment]
@cbrnr 
Most likely, distro package maintainers manually add a patch with client shadows. 

> probably because it already works

We can't say for sure until you learn the upstream source code.

**Comment by cbrnr** [comment]
I'm pretty sure its in upstream, since I'm using Google Chrome and they don't add any patches AFAICS: https://aur.archlinux.org/packages/google-chrome

**Comment by proninyaroslav** [comment]
Yes, you are right, they use packages compiled directly by the Google team.

**Comment by cbrnr** [comment]
@deepak1556 can you reopen this issue please until it is clear if there's anything that could/should be done?

**Comment by deepak1556** [comment]
The custom titlebar implementation in Electron https://github.com/electron/electron/blob/main/shell/browser/ui/views/opaque_frame_view.cc is pulled from https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/ui/views/frame/opaque_browser_frame_view.h which is the custom titlebar implementation for Chrome. The closest CL I could find related to this was https://chromium-review.googlesource.com/c/chromium/src/+/2901661 and it doesn't enable in OpaqueBrowserView.

I will reopen to confirm the behavior with dPWA on linux

**Comment by cbrnr** [comment]
Thanks @deepak1556, let me know if there is anything I can do to help!

**Comment by vangohao** [comment]
Observing this problem in KDE plasma 6.2.4 with X11. VSCode version 1.96.2

**Comment by proninyaroslav** [comment]
We can do it the way it is done in Intellij IDE: add a border that visually separates the window from the background

![Image](https://github.com/user-attachments/assets/ccd8b5ad-0940-4299-9678-9e10dc35dad0)

**Comment by proninyaroslav** [comment]
The Rounded Window Corners Reborn extension for GNOME seems to workaround this issue:
![Image](https://github.com/user-attachments/assets/4053ce79-c9de-4662-9eb9-c0de69a895cd)

**Comment by cbrnr** [comment]
@proninyaroslav nice, this does work, but unfortunately text gets a little blurry when I enable this extension. It's not as bad as XWayland with fractional scaling, but very noticeable (and for me not usable).

**Comment by cbrnr** [comment]
Seems to be a larger issue with GNOME and fractional scaling: https://github.com/flexagoon/rounded-window-corners/issues/24

**Comment by jelmd** [comment]
After upgrade to 1.98.2 (Ubuntu 22.04) this even happens on normal X11/Xorg. I really hate, if apps do not obey system settings/UI and try to re-invent the wheel and making things worse ...

**Comment by lorenzo-cavazzi** [comment]
BTW this happens on KDE Plasma too (version 6.3)

**Comment by aleeraser** [comment]
Happening on Fedora 41, GNOME 47, VSCode 1.99.3 as well.  
It makes it impossible to stack two or more VSCode windows on top of each other.

**Comment by sedax90** [comment]
@deepak1556 any news on this issue? it's very annoying

**Comment by WindustH** [comment]
meet the same problem

**Comment by rubin55** [comment]
FWIW, I'm still running into this with 1.103 from yesterday. I haven't seen any response from MS devs here, this is for sure a very annoying issue; Is there a way this issue can be given more attention?

**Comment by cbrnr** [comment]
See also #247054, and as noted there, this issue is not exclusive to GNOME, it also happens on KDE.

**Comment by shafiqimtiaz** [comment]
Tested on Ubuntu 24.04.3 LTS
Problem persists

**Comment by CJTS15** [comment]
Hello, I saw this thread and was wondering if this is related to this issue. I am on Fedora 43 Workstation Gnome 49. And my VSCode and Cursor is acting like this. Same specs as the above users too.

<img width="1284" height="902" alt="Image" src="https://github.com/user-attachments/assets/59998d4e-5eaa-4abf-b736-df38bbc44247" />

**Comment by otjura** [comment]
Upgraded system from 43 to 44, switched to KDE. Still no shadows or borders, VSCode v1.120.0
 
- Operating System: Fedora Linux 44
- KDE Plasma Version: 6.6.4
- KDE Frameworks Version: 6.26.0
- Qt Version: 6.10.3
- Kernel Version: 7.0.4-200.fc44.x86_64 (64-bit)
- Graphics Platform: Wayland

Additionally my previously reported hack `code --ozone-platform=x11` no longer works on this system to bring the borders back. Additionally Antigravity has lost its borders now too at some point.

**Comment by cmagina** [comment]
Found a workaround for KDE if you use the Breeze window decoration. Open Settings and make a few changes:

1) Colors & Themes -> Window Decorations -> edit Breeze (Pencil button in the bottom right corner of the theme) -> Window-Specific Overrides tab
Add, click 'Detect Window Properties' and click the code window with the cross-hair or just type `code` in the match box.
Enable the "Hide window title bar" option.
OK, OK

2) Window Management -> Window Rules
Add New, click 'Detect Window Properties' and click the code window with the cross-hair or just type `code` in the Window class (application) regex box.
Add Property, select 'No titlebar and frame', leave with defaults.
Apply

Open Code or restart it and you should get a window with the custom titlebar and a window shadow. This works for other Electron based apps like Discord as well.

**Comment by cbrnr** [comment]
Thanks @cmagina, this really works! Coincidentally, I found the following Reddit post (which I haven't tried yet): https://www.reddit.com/r/kde/comments/1tc8454/how_to_get_rounded_corners_shadow_and_outline_on/

It looks like it does some additional things to the borders, maybe also worth a try.
